src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -24,10 +24,293 @@ ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
+    uct_tcp_sockcm_priv_data_hdr_t *hdr = (uct_tcp_sockcm_priv_data_hdr_t *)
+                                          cep->comm_ctx.buf;
+    uct_cm_remote_data_t remote_data;
+    uct_device_addr_t    *dev_addr;
+    size_t               addr_length;
+    ucs_status_t         status;
+
+    /* get the local interface name associated with the connected fd */
+    status = ucs_sockaddr_get_ifname(cep->fd, ifname_str, UCT_DEVICE_NAME_MAX);
+    if (UCS_OK != status) {
+        return status;
+    }
+
+    /* get the device address of the remote peer associated with the connected fd */
+    status = uct_tcp_sockcm_ep_fd_to_remote_dev_addr(cep->fd, &dev_addr, &addr_length);
+    if (UCS_OK != status) {
+        return status;
+    }
+
+    remote_data.field_mask            = UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR        |
+                                        UCT_CM_REMOTE_DATA_FIELD_DEV_ADDR_LENGTH |
+                                        UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA  |
+                                        UCT_CM_REMOTE_DATA_FIELD_CONN_PRIV_DATA_LENGTH;
+    remote_data.dev_addr              = dev_addr;
+    remote_data.dev_addr_length       = addr_length;
+    remote_data.conn_priv_data        = hdr + 1;
+    remote_data.conn_priv_data_length = hdr->length;
+
+    /* the endpoint, passed as the conn_request to the callback, will be passed
+     * to uct_ep_create() which will be invoked by the user and therefore moving
+     * over to its responsibility. */
+    ucs_list_del(&cep->list);
+    cep->listener->conn_request_cb(&cep->listener->super, cep->listener->user_data,
+                                   ifname_str, cep, &remote_data);
+
+    ucs_free(dev_addr);
+    return UCS_OK;
+}
+
+ucs_status_t uct_tcp_sockcm_ep_handle_data_received(uct_tcp_sockcm_ep_t *cep)
+{
+    ucs_status_t status;
+
+    cep->state |= UCT_TCP_SOCKCM_EP_DATA_RECEIVED;
+    uct_tcp_sockcm_ep_init_comm_ctx(cep);
+
+    status = uct_tcp_sockcm_ep_server_invoke_conn_req_cb(cep);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    status = ucs_async_modify_handler(cep->fd, UCS_EVENT_SET_EVWRITE);
+    if (status != UCS_OK) {
+        ucs_error("failed to modify %d event handler to UCS_EVENT_SET_EVWRITE: %s",
+                  cep->fd, ucs_status_string(status));
+    }
+
+out:
+    return status;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_recv_nb(uct_tcp_sockcm_ep_t *cep)
+{
+    size_t recv_length;
+    ucs_status_t status;
+
+    recv_length = uct_tcp_sockcm_ep_get_cm(cep)->priv_data_len - cep->comm_ctx.offset;
+    status = ucs_socket_recv_nb(cep->fd, UCS_PTR_BYTE_OFFSET(cep->comm_ctx.buf,
+                                                             cep->comm_ctx.offset),
+                                &recv_length, NULL, NULL);
+    if ((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS)) {
+        if ((status == UCS_ERR_CANCELED) && (recv_length == 0)) {
+            uct_tcp_sockcm_ep_handle_disconnect(cep, status);
```

Comment:
is it correct to check `recv_length` if `status != UCS_OK`? if yes, why we can't handle if `recv_length > 0`?

Reply:
yes, the status maybe UCS_ERR_NO_PROGRESS and then recv_length is 0.
@dmitrygx right?

Reply:
1) you check NO_PROGRESS line above
2) if NO_PROGRESS, checking recv_length meaningless since "no progress" :)
3) if it's correct, for example for "cancel" in the middle, @dmitrygx pls document it for `ucs_socket_recv_nb` because this is unusually behavior
4) why we can't handle disconnect if recv_length > 0?

Reply:
if UCS_ERR_NO_PROGRESS is returned, then the offset is increased by 0. why is this wrong? i don't think we need to create a new 'if' branch for that.
recv_length is checked only if UCS_ERR_CANCELED is returned.
UCS_ERR_CANCELED is returned from ucs_socket_recv_nb() only if the recv_length is 0.
the indication of a remote disconnect is that the recv function returns successfully but with the recv_length set to 0; in the ucx case, UCS_ERR_CANCELED is returned with recv_length set to 0.

Reply:
@dmitrygx It seems that the ucs_socket_send_nb will never return UCS_ERR_CANCELED since the return value from send() cannot be 0 - https://linux.die.net/man/3/send 
WDYT?

Reply:
@evgeny-leksikov @alinask thanks, will check and fix